### PR TITLE
refactor(warehouse): performance improvements

### PR DESCRIPTION
**Pull request by snarkychef**
> This includes the following changes:
> 1. Improves resource utilisation by distributing different warehouse connection uploads over time.
> 2. Use exists instead of count(*) which is more performant.
> 3. One single query to populate load files table instead querying separately.

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
## Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added unit tests for the code
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

**Comment by achettyiitr** [comment]
@achettyiitr

File: warehouse/warehouse.go [hunk 1]
```diff
@@ -513,22 +515,23 @@ func uploadFrequencyExceeded(warehouse warehouseutils.WarehouseT, syncFrequency
 	return false
 }
 
-func setLastProcessedMarker(warehouse warehouseutils.WarehouseT) {
+func setLastProcessedMarker(warehouse warehouseutils.WarehouseT, lastProcessedTime time.Time) {
 	lastProcessedMarkerMapLock.Lock()
 	defer lastProcessedMarkerMapLock.Unlock()
-	lastProcessedMarkerMap[warehouse.Identifier] = time.Now().Unix()
+	lastProcessedMarkerMap[warehouse.Identifier] = lastProcessedTime.Unix()
```
Comment (achettyiitr) [inline warehouse/warehouse.go:521]: Check with @sivashanmukh for this if this needs to be populated during restarts.

File: warehouse/upload.go [hunk 2]
```diff
@@ -1515,16 +1515,17 @@ func (job *UploadJobT) setStagingFilesStatus(stagingFiles []*StagingFileT, statu
 	return
 }
 
-func (job *UploadJobT) hasLoadFiles(tableName string) (bool, error) {
+func (job *UploadJobT) hasLoadFiles(tableName string) (exists bool, err error) {
```
Comment (achettyiitr) [inline warehouse/upload.go:1518]: Can we make it one query instead of querying every time per table.

Comment (achettyiitr) [reply warehouse/upload.go]: Made this change. Now during the start of exporting data stage. We are populating the loadFIlesTableMap and we are using it throughout.

**Review by achettyiitr — COMMENTED** [review]
1. Delay for the Integration test.
2. Test exists changes if possible in the hosted dump. Let's see if we can come up with some numbers here.

**Review by achettyiitr — COMMENTED** [review]
Can we somehow revisit https://github.com/rudderlabs/rudder-server/blob/4d0992ef0b92bbe16823da6e783d7dd53ccaeda7/warehouse/warehouse.go#L917 this again.

**Review by achettyiitr — APPROVED** [review]
LGTM

**Comment by achettyiitr** [comment]
We have added enableJitterForSyncs to provide jitter in the warehouse syncs. By default, we are making it as false.

File: warehouse/warehouse.go [hunk 3]
```diff
@@ -514,22 +518,24 @@ func uploadFrequencyExceeded(warehouse warehouseutils.WarehouseT, syncFrequency
...
-func (wh *HandleT) createUploadJobsFromStagingFiles(warehouse warehouseutils.WarehouseT, whManager manager.ManagerI, stagingFilesList []*StagingFileT, priority int) {
+func (wh *HandleT) createUploadJobsFromStagingFiles(warehouse warehouseutils.WarehouseT, whManager manager.ManagerI, stagingFilesList []*StagingFileT, priority int) time.Time {
 	// count := 0
 	// Process staging files in batches of stagingFilesBatchSize
 	// Eg. If there are 1000 pending staging files and stagingFilesBatchSize is 100,
 	// Then we create 10 new entries in wh_uploads table each with 100 staging files
 	var stagingFilesInUpload []*StagingFileT
 	var counter int
 	uploadTriggered := isUploadTriggered(warehouse)
+	uploadStartAfter := getUploadStartAfterTime()
```
Comment (abhimanyubabbar) [inline warehouse/warehouse.go:535]: Why not pass it as an argument to the function ?
The reason I am saying it is because it makes the function definition ambiguous and hard to follow.

File: warehouse/snowflake/snowflake.go [hunk 4]
```diff
@@ -161,15 +161,13 @@ func (sf *HandleT) columnExists(columnName string, tableName string) (exists boo
 	return
 }
 
-func (sf *HandleT) schemaExists(schemaname string) (exists bool, err error) {
-	var count int
-	sqlStatement := fmt.Sprintf(`SELECT count(*) FROM INFORMATION_SCHEMA.SCHEMATA WHERE SCHEMA_NAME = '%s'`, sf.Namespace)
-	err = sf.Db.QueryRow(sqlStatement).Scan(&count)
+func (sf *HandleT) schemaExists() (exists bool, err error) {
+	sqlStatement := fmt.Sprintf(`SELECT EXISTS ( select 1 FROM INFORMATION_SCHEMA.SCHEMATA WHERE SCHEMA_NAME = '%s' )`, sf.Namespace)
+	err = sf.Db.QueryRow(sqlStatement).Scan(&exists)
```
[the remark this answers is not in the document]
Comment (achettyiitr) [reply warehouse/snowflake/snowflake.go:166]: Done.

**Review by sivashanmukh — APPROVED** [review]
LGTM
